File: plugins/robots/robotsGenerator/nxtOSEK/generators/simpleElements/waitForColorIntensityBlockGenerator.cpp
#include "waitForColorIntensityBlockGenerator.h"
#include "../../nxtOSEKRobotGenerator.h"
#include "../listsmartline.h"

using namespace robots::generator;

WaitForColorIntensityBlockGenerator::WaitForColorIntensityBlockGenerator()
{
}

void WaitForColorIntensityBlockGenerator::addInitAndTerminateCode(NxtOSEKRobotGenerator *nxtGen
		, QString const &port, qReal::Id const &elementId)
{
	QString const partInitCode = "ecrobot_init_nxtcolorsensor(" + port;
	if (!ListSmartLine::isContainsPart(nxtGen->initCode(), partInitCode)) {
		// i don't no, mb NXT_LIGHTSENSOR_NONE, NXT_COLORSENSOR_DEACTIVATE
		QString const initCode = "ecrobot_init_nxtcolorsensor(" + port + "," + "NXT_COLORSENSOR);";
		QString const terminateCode = "ecrobot_term_nxtcolorsensor(" + port + ");";
		nxtGen->initCode().append(SmartLine_old(initCode, elementId));
		nxtGen->terminateCode().append(SmartLine_old(terminateCode, elementId));
	}
}

QList<SmartLine_old> WaitForColorIntensityBlockGenerator::convertElementIntoDirectCommand(NxtOSEKRobotGenerator *nxtGen
		, qReal::Id const &elementId, qReal::Id const &logicElementId)
{
	QList<SmartLine_old> result;
	QString portStr = nxtGen->api()->stringProperty(logicElementId, "Port");
	QString const intensity = nxtGen->api()->stringProperty(logicElementId,  "Intensity");
	QString const inequalitySign = transformSign(QString(nxtGen->api()->stringProperty(logicElementId
			, "Sign").toUtf8()));

	QString const condition = inequalitySign + " " + intensity;

	result.append(SmartLine_old("while (!(ecrobot_get_nxtcolorsensor_light(NXT_PORT_S" + portStr
			+ ") * 100 / 1023 " + condition + ")) {", elementId));
	result.append(SmartLine_old("}", elementId));

	addInitAndTerminateCode(nxtGen, "NXT_PORT_S" + portStr, elementId);

	return result;
}

File: plugins/tools/visualInterpreter/examples/robotsCodeGeneration/reactionsStorage/WaitForColorIntensityBlockGenerator.py
# Application condition
waitFor.id == max_used_id and not cur_node_is_processed

# Reaction
port = "NXT_PORT_S" + waitFor.Port
condition = convertCondition(waitFor.Sign) + " " + waitFor.Intensity

wait_for_color_intensity_code = "while (!(ecrobot_get_nxtcolorsensor_light(" + port + ") " + condition + ")) {}\n"
wait_init_code = "ecrobot_init_nxtcolorsensor(" + port + ", NXT_COLORSENSOR));\n"
wait_terminate_code = "ecrobot_term_nxtcolorsensor(" + port + ");\n"

if wait_init_code not in init_code:
  init_code.append(wait_init_code)
  terminate_code.append(wait_terminate_code)

code.append([wait_for_color_intensity_code])
id_to_pos_in_code[waitFor.id] = len(code) - 1

cur_node_is_processed = True
